### PR TITLE
Add support for sharing an ORT session

### PR DESCRIPTION
**Pull request by quic-suppugun**
For every instance in a model instance group a new ORT session is
created. This code adds support to share a session per instance
group.
This support can be enabled by defining 'share_session' to true
in triton model config "parameters". Example:
parameters [
.....
  {
    key: "share_session"
    value: {string_value: "true"}
  }
]

This is a global parameter and cannot be defined per instance
group. The user should determine if the parameter makes sense for
their setup. 
GetInstanceGroupName function is added to find the instance group 
name through regex search over instance name.

### CODE REVIEW

**Comment by mleies** [comment]
Thanks @quic-suppugun  for this!

Sharing  model weights across instances is a huge benefit for us as we use multiple instances to improve throughput, but with larger models memory becomes the bottle neck.  We were going to implement a similar fix when we came across this PR.  

Please merge:
- We are currently using this fix, merged into 22.07, without issue
- It is of little risk, as the fix is only used when model config files specify sharing sessions

**Comment by dyastremsky** [comment]
Thank you for this contribution! We'll need you to submit a signed CLA (see [here](https://github.com/triton-inference-server/server/blob/main/CONTRIBUTING.md)) to review this. Please let me know once you have sent it in.

CC: @pranavsharma

**Comment by mleies** [comment]
@dyastremsky, 

Getting the CLA sorted will keep you posted.   

We've now merged this PR successfully into 23.06, with minor changes.   I plan to review and add those changes, assuming the github setup will let me modify this PR.    What's the most expedient process?

**Comment by dyastremsky** [comment]
Thank you, @mleies. The most expedient way would be if you also added a test to the server QA folder [here](https://github.com/triton-inference-server/server/tree/main/qa), or perhaps added to the L0_onnx_optimization test if it's a fit. If not, we'd need to get someone internally to add the testing. We'd also run CI to ensure nothing breaks.

Ideally, @pranavsharma would review the PR once it's ready with no conflicts. If Pranav does not have time, I can also review it.

That review process starts with getting the CLA in though, since we cannot merge your changes without a CLA.

**Comment by dyastremsky** [comment]
Also, @mleies, not sure if you and @quic-suppugun are on the same team. If not, we'd need to close out this PR and need a fresh PR with a CLA provided by whoever wrote the code.

File: src/onnxruntime_utils.cc [hunk 1]
```diff
@@ -493,5 +493,22 @@ CompareDimsSupported(
...
+    const std::string& model_name, const std::string& instance_name)
+{
+  std::regex groupNameRegex('(' + model_name + '_' + "[0-9]" + ')');
+  std::smatch groupName;
+
+  if (model_name.empty() || instance_name.empty()) {
+    return "";
+  }
+
+  if (std::regex_search(instance_name, groupName, groupNameRegex)) {
```
Comment (pranavsharma) [inline src/onnxruntime_utils.cc:507]: Can you explain and document why a regex search is needed?

File: src/onnxruntime.cc [hunk 2]
```diff
@@ -881,6 +932,38 @@ ModelState::AutoCompleteIO(const char* key, const OnnxTensorInfoMap& io_infos)
   return nullptr;  // success
 }
 
+TRITONSERVER_Error*
+ModelState::GetSessionForGroup(
+    const std::string& group_name, std::shared_ptr<OrtSession>& session)
+{
+  RETURN_ERROR_IF_TRUE(
+      group_name.empty(), TRITONSERVER_ERROR_INVALID_ARG,
+      std::string("Invalid group name"));
```
Comment (pranavsharma) [inline src/onnxruntime.cc:941]: Please print the name of the group in the error msg.

File: src/onnxruntime.cc [hunk 3]
```diff
@@ -881,6 +932,38 @@ ModelState::AutoCompleteIO(const char* key, const OnnxTensorInfoMap& io_infos)
...
+  {
+    std::unordered_map<std::string, std::shared_ptr<OrtSession>>::iterator
+        sessionEntry;
+    sessionEntry = groupInstanceSessionMap_.find(group_name);
+    RETURN_ERROR_IF_TRUE(
+        (sessionEntry == groupInstanceSessionMap_.end()),
+        TRITONSERVER_ERROR_NOT_FOUND, std::string("No such group"));
```
Comment (pranavsharma) [inline src/onnxruntime.cc:948]: Please print the name of the group in the error msg.

File: src/onnxruntime.cc [hunk 4]
```diff
@@ -881,6 +932,38 @@ ModelState::AutoCompleteIO(const char* key, const OnnxTensorInfoMap& io_infos)
...
+  return nullptr;
+}
+
+TRITONSERVER_Error*
+ModelState::SetSessionForGroup(
+    const std::string& group_name, const std::shared_ptr<OrtSession>& session)
+{
+  RETURN_ERROR_IF_TRUE(
+      group_name.empty(), TRITONSERVER_ERROR_INVALID_ARG,
+      std::string("Invalid group name"));
```
Comment (pranavsharma) [inline src/onnxruntime.cc:961]: Please print the name of the group in the error msg.

File: src/onnxruntime.cc [hunk 5]
```diff
@@ -373,6 +403,20 @@ ModelState::LoadModel(
     }
   }
 
+  // Check is we are sharing the session. If so get the session pointer and
+  // return
+  if (share_session_) {
+    if (GetSessionForGroup(instance_group_name, session) == nullptr) {
```
Comment (pranavsharma) [inline src/onnxruntime.cc:409]: Is there a potential for a race condition where multiple instances are coming up and each end up creating a Session? If ModelState is always initialized before the instances, there's no issue and we won't need to use a lock. (I don't quite remember the sequence).

File: src/onnxruntime.cc [hunk 6]
```diff
@@ -110,6 +114,17 @@ class ModelState : public BackendModel {
   // is specified both in the output section and state section, it indicates
   // that the backend must return the output state to the client too.
   std::map<std::string, std::pair<int64_t, int64_t>> model_outputs_;
+
+  // Indicate if an onnxrt session should be shared or not. This is a model
+  // global and applies to all instances. So, storing it in the model state
+  bool share_session_;
```
Comment (pranavsharma) [inline src/onnxruntime.cc:120]: Would be nice to make the config more descriptive like share_session_between_instances.

**Review by pranavsharma — COMMENTED** [review]
Is there a PR for documentation as well?

**Comment by Jackiexiao** [comment]
Any update? this PR is great, can't wait to see it merge

**Comment by Jackiexiao** [comment]
Hello @pranavsharma, @mleies, and @dyastremsky,

I am reaching out in hopes of facilitating a swift merge for the current PR. I've constructed an ONNX image based on the 23.12 build (see [mleies's docs](https://github.com/mleies/onnxruntime_backend/blob/backend_w_session_sharing/shared_session.md)) , which you can find below:

Code repository: https://github.com/Jackiexiao/onnxruntime_backend/tree/r23.12
Docker image: jackiexiao/tritonserver:23.12-onnx-py-cpu-share-session (available for pull)

Additionally, I've updated the main branch at https://github.com/Jackiexiao/onnxruntime_backend/tree/main with the following revisions:

- Incorporated feedback from @pranavsharma, addressing a portion of the review comments.
- Resolved merge conflicts due to the PR's basis on an earlier version of the codebase.
- As I am not deeply familiar with C++, I have a specific change that I would like to confirm. To ensure the share_session_between_instances feature is operational, I've set ParallelModelInstanceLoading to False. Here is the change:
```
  RETURN_IF_ERROR(TRITONBACKEND_BackendAttributeSetParallelModelInstanceLoading(
      backend_attributes, false));
```
You can review this modification in context here: https://github.com/Jackiexiao/onnxruntime_backend/blob/128f7aa4a3eb4b4ad94f171824e85c48ec6303a3/src/onnxruntime.cc#L2981C1-L2983C35

I would greatly appreciate any assistance or additional feedback to expedite the merging process.

Thank you for your support and guidance.

**Comment by Jackiexiao** [comment]
@tanmayv25 @mc-nv @kthui  Could you please take a look at this Pull Request? Thank you.

**Comment by quic-suppugun** [comment]
@pranavsharma @dyastremsky, I addressed review comments in [PR 248](https://github.com/triton-inference-server/onnxruntime_backend/pull/248).

**Pull request closed without merging**